server.go

```diff
@@ -295,6 +295,15 @@ func (srv *Server) HandleConn(newConn net.Conn) {
 		}
 		return
 	}
+	// If we have a callback, also call it once the connection goes down
+	// this gives people a chance to e.g. log disconnections that happen
+	// after the initial handshake
+	if srv.ConnectionFailedCallback != nil {
+		go func() {
+			wErr := sshConn.Wait()
+			srv.ConnectionFailedCallback(conn, wErr)
```

Comment:
Instead of spawning a goroutine that can theoretically be left running after the server has closed, we could move this check to after the for loop `for ch := range chans { ... }`. That channel will be closed anyway before `.Wait()` here returns.